source/BrightcoveOS .NET-MAPI-Wrapper/Api/BrightcoveApi.video.write.cs

```diff
@@ -430,18 +430,86 @@ private BrightcoveImage DoAddImage(BrightcoveImage image, FileUploadInfo fileUpl
-		public BrightcoveLogoOverlay AddLogoOverlay(BrightcoveLogoOverlay logoOverlay, FileUploadInfo fileUploadInfo, long videoId)
+        #endregion
+
+        #region AddCaptioning
+        public BrightcoveCaptioning AddCaptioning(string captionFileUrl, long videoId, string filename)
+        {
+            return DoAddCaptioningWithExternalUrl(captionFileUrl, videoId, null, filename);
+        }
+
+        public BrightcoveCaptioning AddCaptioning(string captionFileUrl, string videoReferenceId, string filename)
+        {
+            return DoAddCaptioningWithExternalUrl(captionFileUrl, -1, videoReferenceId, filename);
+        }
+
+        private BrightcoveCaptioning DoAddCaptioningWithExternalUrl(string captionFileUrl, long videoId, string videoReferenceId, string filename)
+        {
+            string propName;
+            object propValue;
+            GetIdValuesForUpload(videoId, videoReferenceId, "video_id", "video_reference_id", out propName, out propValue);
+
+            var captionSource = new BrightcoveCaptionSource {
+                Complete = true,
+                IsRemote = true,
+                DisplayName = filename,
+                Url = captionFileUrl
+            };
+
+            BrightcoveParamCollection parms = CreateWriteParamCollection("add_captioning",
+                                                                         methodParams => {
+                                                                             methodParams.Add(propName, propValue);
+                                                                             methodParams.Add("caption_source", captionSource);
+                                                                             methodParams.Add("filename", filename);
+                                                                         });
+
+            return RunPost<BrightcoveResultContainer<BrightcoveCaptioning>>(parms).Result;
+        }
+
+        public BrightcoveCaptioning AddCaptioning(FileUploadInfo captionFileUploadInfo, long videoId, string filename)
+        {
+            return DoAddCaptioningWithFileUpload(captionFileUploadInfo, videoId, null, filename);
+        }
+
+        public BrightcoveCaptioning AddCaptioning(FileUploadInfo captionFileUploadInfo, string videoReferenceId, string filename)
+        {
+            return DoAddCaptioningWithFileUpload(captionFileUploadInfo, -1, videoReferenceId, filename);
+        }
+
+        private BrightcoveCaptioning DoAddCaptioningWithFileUpload(FileUploadInfo captionFileUploadInfo, long videoId, string videoReferenceid, string filename)
+	    {
+            string propName;
+            object propValue;
+            GetIdValuesForUpload(videoId, videoReferenceid, "video_id", "video_reference_id", out propName, out propValue);
+
+            var captionSource = new BrightcoveCaptionSource {
+                Complete = true,
+                IsRemote = false,
+                DisplayName = filename,

```

Comment:
I left it as their API docs said. We can change it to use what suits us in our code.